S4.Rmd

```diff
@@ -143,6 +158,16 @@ As a user, you should generally not extract values out of someone else's object
 
 We can override the default behaviour by providing an `initialize()` method. You'll learn more about the details of creating methods in the next section, but in brief, the following code defines a `Person` method for the `initialize` generic.
 
+<!-- FIXME: initialize has a complicated contract that is not easily satisfied; most experienced Bioconductor developers don't actually write initialize methods, but rather a constructor Person() that checks / etc and then calls new() (like Factor(), below). For instance, S4Vectors defines 35 classes but no initialize methods, IRanges has 85 classes and no initialize methods.
```

Comment:
Since unnamed arguments to (the default) `initialize()` are taken as base classes for initialization, one pattern is

```
.Employee <- setClass(
    "Employee",
    contains = "Person",
    slots = c(
        job_title = "character",
        id = "character"
    )
)

Employee <-
    function(name = character(), age = numeric(), job_title = character())
{
    stopifnot(length(name) == length(job_title))

    id <- paste0("EID-", sample(100000, length(job_title))) # psuedo-unique
    .Employee(Person(name, age), job_title = job_title, id = id)
}
```